packages/eslint-plugin/src/rules/prefer-for-of.ts

```diff
@@ -0,0 +1,198 @@
+    ): TSESTree.Expression | null {
+      let arrayExpression: TSESTree.Expression | null = null;
+      if (
+        node !== null &&
+        node.type === AST_NODE_TYPES.BinaryExpression &&
+        node.operator === '<' &&
+        isMatchingIdentifier(node.left, name) &&
+        node.right.type === AST_NODE_TYPES.MemberExpression &&
+        isMatchingIdentifier(node.right.property, 'length')
+      ) {
+        arrayExpression = node.right.object;
+      }
+      return arrayExpression;
+    }
+
+    function isIncrement(node: TSESTree.Node | null, name: string): boolean {
+      if (!node) {
+        return false;
+      }
+      switch (node.type) {
+        case AST_NODE_TYPES.UpdateExpression:
+          // x++ or ++x
+          return (
+            node.operator === '++' && isMatchingIdentifier(node.argument, name)
+          );
+        case AST_NODE_TYPES.AssignmentExpression:
+          if (isMatchingIdentifier(node.left, name)) {
+            if (node.operator === '+=') {
+              // x += 1
+              return isLiteral(node.right, 1);
+            } else if (node.operator === '=') {
+              // x = x + 1 or x = 1 + x
+              const expr = node.right;
+              return (
+                expr.type === AST_NODE_TYPES.BinaryExpression &&
+                expr.operator === '+' &&
+                ((isMatchingIdentifier(expr.left, name) &&
+                  isLiteral(expr.right, 1)) ||
+                  (isLiteral(expr.left, 1) &&
+                    isMatchingIdentifier(expr.right, name)))
+              );
+            }
+          }
+      }
+      return false;
+    }
+
+    function contains(outer: TSESTree.Node, inner: TSESTree.Node): boolean {
+      return (
+        outer.range[0] <= inner.range[0] && outer.range[1] >= inner.range[1]
+      );
+    }
+
+    function isAssignee(node: TSESTree.Node): boolean {
+      const parent = node.parent;
+      if (!parent) {
+        return false;
+      }
+      return (
+        // a[i] = 1, a[i] += 1, etc.
+        (parent.type === AST_NODE_TYPES.AssignmentExpression &&
+          parent.left === node) ||
+        // delete a[i]
+        (parent.type === AST_NODE_TYPES.UnaryExpression &&
+          parent.operator === 'delete' &&
+          parent.argument === node) ||
+        // a[i]++, --a[i], etc.
+        (parent.type === AST_NODE_TYPES.UpdateExpression &&
+          parent.argument === node) ||
+        // [a[i]] = [0]
+        parent.type === AST_NODE_TYPES.ArrayPattern ||
+        // [...a[i]] = [0]
+        parent.type === AST_NODE_TYPES.RestElement ||
+        // ({ foo: a[i] }) = { foo: 0 }
+        (parent.type === AST_NODE_TYPES.Property &&
+          parent.parent !== undefined &&
+          parent.parent.type === AST_NODE_TYPES.ObjectExpression &&
+          parent.value === node &&
+          isAssignee(parent.parent))
+      );
+    }
+
+    function isIndexOnlyUsedWithArray(
+      context: RuleContext<'preferForOf', never[]>,
```

Comment:
you don't need to pass the context into the function, it's available in the function's parent closure.
```suggestion
```